In [ ]:
!git clone https://github.com/sanjivinicarmel/p53-Mutation-Using-ML-Techniques.git

Cloning into 'p53-Mutation-Using-ML-Techniques'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (3/3), done.


In [ ]:
%cd p53-Mutation-Using-ML-Techniques

/content/p53-Mutation-Using-ML-Techniques


In [ ]:
import os

os.environ['GITHUB_TOKEN'] = "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"
!git config --global user.email "sanjivinicarmel2001@gmail.com"
!git config --global user.name "sanjivinicarmel"


Understanding p53 Mutations: Predicting How Gene Changes Affect Protein **Function**

*This project aims to predict how specific changes (mutations) in the p53 protein affect its function, such as losing normal activity, gaining new harmful effects, or interfering with the normal protein. By analyzing different characteristics of these mutations, we can classify their impact using machine learning techniques. This helps in understanding how mutations might contribute to diseases like cancer.*

Table of Contents:


1.  About the dataset
2.  Loading the dataset Preprocessing
3. Explanatory Data Analysis



1. About the Dataset


1.   Mutation Details: Includes specific amino acid changes (e.g., V143A), the position of the mutation (Codon_Number), and the original and mutated amino acids (WT_AA and Mutant_AA).
2.   Structural & Functional Context: Information on which structural motifs or domains are affected (like L2/L3), and conserved wild-type functions such as DNA binding.
3. Experimental Conditions: Data on assay temperature, cell types and lines used, and experimental design details.
4. Methodology: Types of experimental methods applied (e.g., Luciferase assay).
5. Target Variables: Categorical labels indicating the mutation’s effect, such as Loss_of_Function, Dominant_Negative_Activity, or Gain_of_Function.
6. Size and Scope: Contains thousands of mutation entries, each described by these multiple features, providing a rich dataset for training accurate predictive models.



2. Loading the dataset

In [ ]:
import pandas as pd


In [ ]:
mutation_raw_data=pd.read_csv(r"/content/FunctionDownload_r20.csv")
mutation_raw_data.head(5)

Function_ID ProtDescription AAchange  Codon_Number WT_AA Mutant_AA  \
0            1         p.V143A    V143A           143   Val       Ala   
1            2         p.R175H    R175H           175   Arg       His   
2            6         p.R248W    R248W           248   Arg       Trp   
3            7         p.R273H    R273H           273   Arg       His   
4            8         p.D281G    D281G           281   Asp       Gly   

   SwissProtLink  Structural_motif Codon72AA Conserved_WT_Function  ...  \
0         5887.0  NDBL/beta-sheets         R         DNAb (p53CON)  ...   
1         5932.0             L2/L3       NaN         DNAb (p53CON)  ...   
2         5984.0             L2/L3       NaN                   NaN  ...   
3         5995.0           L1/S/H2       NaN         DNAb (p53CON)  ...   
4         6012.0           L1/S/H2       NaN                   NaN  ...   

  Assay_design                  Method FRef_ID  \
0   WT control  Luciferase assay, EMSA       1   
1   WT control  Luciferase assay, EMSA       1   
2   WT control  Luciferase assay, EMSA       1   
3   WT control  Luciferase assay, EMSA       1   
4   WT control  Luciferase assay, EMSA       1   

                                            Authors  \
0  Zhang W;Funk WD;Wright WE;Shay JW;Deisseroth AB;   
1  Zhang W;Funk WD;Wright WE;Shay JW;Deisseroth AB;   
2  Zhang W;Funk WD;Wright WE;Shay JW;Deisseroth AB;   
3  Zhang W;Funk WD;Wright WE;Shay JW;Deisseroth AB;   
4  Zhang W;Funk WD;Wright WE;Shay JW;Deisseroth AB;   

                                               Title  Year   Journal Volume  \
0  Novel DNA binding of p53 mutants and their rol...  1993  Oncogene    8.0   
1  Novel DNA binding of p53 mutants and their rol...  1993  Oncogene    8.0   
2  Novel DNA binding of p53 mutants and their rol...  1993  Oncogene    8.0   
3  Novel DNA binding of p53 mutants and their rol...  1993  Oncogene    8.0   
4  Novel DNA binding of p53 mutants and their rol...  1993  Oncogene    8.0   

  Start_page   PubMed  
0     2555.0  8361764  
1     2555.0  8361764  
2     2555.0  8361764  
3     2555.0  8361764  
4     2555.0  8361764  

[5 rows x 27 columns]

In [ ]:
mutation_data_clone=mutation_raw_data.copy()

In [ ]:
#Finding the total amount of data
mutation_data_clone.shape

(7570, 27)

In [ ]:
mutation_data_clone.duplicated().sum()

np.int64(0)

No Duplicate Record Found

In [ ]:
mutation_data_clone.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7570 entries, 0 to 7569
Data columns (total 27 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Function_ID                 7570 non-null   int64  
 1   ProtDescription             7570 non-null   object 
 2   AAchange                    7570 non-null   object 
 3   Codon_Number                7570 non-null   int64  
 4   WT_AA                       7570 non-null   object 
 5   Mutant_AA                   6640 non-null   object 
 6   SwissProtLink               3939 non-null   float64
 7   Structural_motif            7570 non-null   object 
 8   Codon72AA                   879 non-null    object 
 9   Conserved_WT_Function       3223 non-null   object 
 10  Loss_of_Function            4412 non-null   object 
 11  Dominant_Negative_Activity  652 non-null    object 
 12  Gain_of_Function            709 non-null    object 
 13  Temperature_Sensitivity     789 n

We observe the presence of null values in close to 6 columns , which will be dealt with in upcoming steps .

In [ ]:
mutation_data_clone.describe ()

Function_ID  Codon_Number  SwissProtLink      FRef_ID         Year  \
count  7570.000000   7570.000000    3939.000000  7570.000000  7570.000000   
mean   4674.033818    208.132761   26334.412541   249.856275  2010.390489   
std    2536.875243     63.130061   19363.686369   125.024628     8.571089   
min       1.000000      1.000000    5853.000000     1.000000  1990.000000   
25%    2753.250000    158.000000    5975.000000   115.000000  2003.000000   
50%    4941.500000    211.000000   36509.000000   347.000000  2018.000000   
75%    6833.750000    257.000000   45103.500000   347.000000  2018.000000   
max    8726.000000    392.000000   47214.000000   351.000000  2019.000000   

            Volume    Start_page        PubMed  
count  7357.000000   7242.000000  7.570000e+03  
mean     67.836210   1472.584093  2.193051e+07  
std      61.132697   3042.675705  8.835777e+06  
min       1.000000      1.000000  1.322785e+06  
25%      24.000000    178.000000  1.290972e+07  
50%      71.000000    178.000000  2.997996e+07  
75%      71.000000   1369.000000  2.997996e+07  
max     581.000000  39359.000000  3.088612e+07

1. Function_ID	: Ranges from 1 to 8726, likely representing different samples or mutation types.
2. Codon_Number :	Codon positions range from 1 to 392 — consistent with the p53  gene's coding region (~393 codons).

3. The mean codon number (~208) shows that:
       

*    Mutations in your dataset are not random,
*   They are concentrated around the central, functionally important region of the p53 protein — especially the DNA-binding domain, where mutations are most likely to disrupt tumor suppressor function.

4. The other columns will be dropped because they will refer to book publication and their years.





**Percentage of missing values**

In [ ]:
missing_data_ratio= (mutation_data_clone.isnull().sum()/len(mutation_data_clone)*100)
missing_data_ratio=missing_data_ratio[missing_data_ratio>0]
print(missing_data_ratio)

Mutant_AA                     12.285337
SwissProtLink                 47.965654
Codon72AA                     88.388375
Conserved_WT_Function         57.424042
Loss_of_Function              41.717305
Dominant_Negative_Activity    91.387054
Gain_of_Function              90.634082
Temperature_Sensitivity       89.577279
Temp_ref                       6.671070
Cell_lines                     0.013210
Assay_design                   0.013210
Method                         0.079260
Volume                         2.813738
Start_page                     4.332893
dtype: float64


A missing data analysis revealed that approximately 7 out of 14 columns have more than 50% missing values, with some exceeding 90%. This pattern suggests either incomplete data collection or limited experimental validation in the source database. Features such as Cell_lines and Assay_design, however, have near-complete information, making them reliable for modeling.

**Droping irrelevant Columns**

In [ ]:
mutation_data_clone.drop(columns=["Function_ID", "ProtDescription", "SwissProtLink", "FRef_ID", "Authors", "Title", "Year", "Journal", "Volume", "Start_page", "PubMed"], inplace=True)
#Number of columns reduced from 27 to 16

**One hot for handling categorical data [WT_AA", "Mutant_AA", "Structural_motif", "Method]**

In [ ]:
categorical_cols = ["WT_AA", "Mutant_AA", "Structural_motif", "Method"]
mutation_data_clone = pd.get_dummies(mutation_data_clone, columns=categorical_cols)
mutation_data_clone

AAchange  Codon_Number Codon72AA  \
0       V143A           143         R   
1       R175H           175       NaN   
2       R248W           248       NaN   
3       R273H           273       NaN   
4       D281G           281       NaN   
...       ...           ...       ...   
7565    G360V           360       NaN   
7566    R248Q           248       NaN   
7567    R273C           273       NaN   
7568   P359fs           359       NaN   
7569   G361fs           361       NaN   

                                  Conserved_WT_Function  \
0                                         DNAb (p53CON)   
1                                         DNAb (p53CON)   
2                                                   NaN   
3                                         DNAb (p53CON)   
4                                                   NaN   
...                                                 ...   
7565  SuperGS; superAPO; TA (RGC, BAX, WAF1) supertrans   
7566                                                NaN   
7567                                                NaN   
7568                               TA (WAF1) supertrans   
7569                               TA (WAF1) supertrans   

                     Loss_of_Function Dominant_Negative_Activity  \
0       TA (p53CON) partial; TA (RGC)                        NaN   
1                    TA (p53CON, RGC)                        NaN   
2     DNAb (p53CON); TA (p53CON, RGC)                        NaN   
3       TA (p53CON) partial; TA (RGC)                        NaN   
4     DNAb (p53CON); TA (p53CON, RGC)                        NaN   
...                               ...                        ...   
7565                              NaN                        NaN   
7566                              NaN                        NaN   
7567                              NaN                        NaN   
7568                              NaN                        NaN   
7569                              NaN                        NaN   

     Gain_of_Function Temperature_Sensitivity Temp_ref Cell_assay  ...  \
0                 NaN                     NaN     37°C      Human  ...   
1                 NaN                     NaN     37°C      Human  ...   
2                 NaN                     NaN     37°C      Human  ...   
3                 NaN                     NaN     37°C      Human  ...   
4                 NaN                     NaN     37°C      Human  ...   
...               ...                     ...      ...        ...  ...   
7565              NaN                     NaN     37°C      Human  ...   
7566        TA (HER2)                     NaN     37°C      Human  ...   
7567        TA (HER2)                     NaN     37°C      Human  ...   
7568              NaN                     NaN     37°C      Human  ...   
7569              NaN                     NaN     37°C      Human  ...   

     Method_siRNA and oxaliplatin or cisplatin treatment  \
0                                                 False    
1                                                 False    
2                                                 False    
3                                                 False    
4                                                 False    
...                                                 ...    
7565                                              False    
7566                                              False    
7567                                              False    
7568                                              False    
7569                                              False    

     Method_siRNA and western blot  Method_siRNA, RT-PCR  \
0                            False                 False   
1                            False                 False   
2                            False                 False   
3                            False                 False   
4                            False                 False   
...             

**Filling missing values with unknown**

In [ ]:
# Handling Missing Values
mutation_data_clone.fillna({"Mutant_AA": "Unknown", "Codon72AA": "Unknown", "Conserved_WT_Function": "Unknown"}, inplace=True)

1. Preserves Data Integrity: Filling with "Unknown" retains all rows and avoids dropping potentially valuable mutation data.

2. Makes Missingness Explicit: Clearly indicates that information is unavailable, rather than guessing or misrepresenting it.

3. Model-Friendly Encoding: Treats missing values as a distinct category, allowing machine learning models to potentially learn from the absence of information.

In [ ]:
!git status

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [ ]:
!git add .


In [ ]:
!git commit -m "Initial Cleaning done "


On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [ ]:
!git push https://sanjivinicarmel:$GITHUB_TOKEN@github.com/sanjivinicarmel/p53-mutation-Using-ML-Techniques.git

Everything up-to-date
